sphinx/content/ch1.md

```diff
@@ -0,0 +1,1249 @@
+
+    RPKM = (10^9 * C) / (N * L)
+
+    Where:
+    C = Number of reads mapped to a gene
+    N = Total mapped reads in the experiment
+    L = Exon length in base pairs for a gene
+
+    Parameters
+    ----------
+    counts: array, shape (N_genes, N_samples)
+        RNAseq (or similar) count data where columns are individual samples
+        and rows are genes.
+    lengths: array, shape (N_genes,)
+        Gene lengths in base pairs in the same order
+        as the rows in counts.
+
+    Returns
+    -------
+    normed : array, shape (N_genes, N_samples)
+        The RPKM normalized counts matrix.
+    """
+    # First, convert counts to float to avoid overflow when multiplying by
+    # 1e9 in the RPKM formula
+    C = counts.astype(float)
+    N = np.sum(C, axis=0)  # sum each column to get total reads per sample
+    L = lengths
+
+    normed = 1e9 * C / (N[np.newaxis, :] * L[:, np.newaxis])
+
+    return(normed)
+```
+
+The `rpkm` function defined in {numref}`code:intro` illustrates some of the
+ways that NumPy arrays can make your code more elegant:
+
+ - Arrays can be 1D, like lists, but they can also be 2D, like matrices, and
+   higher-dimensional still. This allows them to represent many different
+   kinds of numerical data. In our case, we are manipulating a 2D matrix.
+ - Arrays can be operated on along *axes*. In the first line, we calculate the
+   sum down each column by specifying `axis=0`.
+ - Arrays allow the expression of many numerical operations at once.
+   For example toward the end of the function we divide the 2D array of counts
+   (C) by the 1D array of column sums (N).
+   This is broadcasting. More on how this works in just a moment!
+
+Before we delve into the power of NumPy, let's spend some time looking at the biological data that we will be working with.
+
+## Introduction to the Data: What Is Gene Expression?
+
+We will work our way through a *gene expression analysis* to demonstrate the power of NumPy and SciPy to solve a real-world biological problem.
+We will use the pandas library, which builds on NumPy, to read and munge our data files, and then we will manipulate our data efficiently in NumPy arrays.
+
+The so-called [central dogma of molecular biology](https://en.wikipedia.org/wiki/Central_dogma_of_molecular_biology) (see {numref}`fig:central_dogma`) states that all the information needed to run a cell (or an organism, for that matter) is stored in a molecule called *deoxyribonucleic acid*, or DNA.
+{numref}`fig:dna` presents a high-level overview of the DNA molecule.
+This molecule has a repetitive backbone on which lie chemical groups called *bases*, in sequence.
+There are four kinds of bases, abbreviated as A, C, G, and T, comprising an alphabet with which information is stored.
+
+```{figure} https://upload.wikimedia.org/wikipedia/commons/e/e4/DNA_chemical_structure.svg
+---
+name: fig:dna
+---
+The chemical structure of DNA (image by Madeleine Price Ball, used under
```

Comment:
Kind of weird that, here, the caption is *not* in the meta-data.

Reply:
Agreed, though that quirk comes directly from [rST](https://docutils.sourceforge.io/docs/ref/rst/directives.html#figure) itself